Download Keyboard dictionaries on demand

HollowMan6 opened the pull request:
Resolve #1128

We remove the SQLiteAssetHelper dependency, store the database into the dedicated database path. Now the auto-complete will be on after the dictionaries are downloaded successfully.

We continue to store Chinese databases locally as those should be available immediately to type in Chinese.

HollowMan6 commented:
To test, replace the value of the build variable PROPS_ENDPOINT in app/build.gradle with "https://raw.githubusercontent.com/Igalia/wolvic/gh-pages-dics/props.json" and recompile.

buildConfigField 'String', 'PROPS_ENDPOINT', '"https://raw.githubusercontent.com/Igalia/wolvic/gh-pages-dics/props.json"'

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionariesManager.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Nullable
+    public String getOrDownloadDictionary(@NonNull String lang) {
```
Looks like the return value is never used. Can we get rid of it? Or where you thinking in other use cases were it could be useful?

HollowMan6 replied:
Let's just remove the return value here.

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionariesManager.java:
```diff
@@ -0,0 +1,153 @@
+
+    private void downloadDictionary(@NonNull String lang) {
+        final Dictionary dictionary = DictionaryUtils.getExternalDictionaryByLang(mContext, lang);
+        final String payload = DictionaryUtils.getDictionaryPayload(dictionary);
```
you're using dictionary here without any check. It can be null

HollowMan6 replied:
Fixed and opened the PR #1175 to fix the same thing in Environments manager.

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionariesManager.java:
```diff
@@ -0,0 +1,153 @@
+    private void downloadDictionary(@NonNull String lang) {
+        final Dictionary dictionary = DictionaryUtils.getExternalDictionaryByLang(mContext, lang);
+        final String payload = DictionaryUtils.getDictionaryPayload(dictionary);
+        if (dictionary != null) {
```
I prefer to have an early return if == null to save some indentation

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionariesManager.java:
```diff
@@ -0,0 +1,153 @@
+                mDownloadManager.removeDownload(dicItem.getId(), true);
+            } else {
+                // The 'downloadDictionary' method is called either by 'getOrDownloadDictionary' when the user changes / type the keyboard
+                Log.w(LOGTAG, "Download is still in progress; we shouldn't reach this code.");
```
Maybe an assert?

HollowMan6 replied:
Same in #1175

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionariesManager.java:
```diff
@@ -0,0 +1,153 @@
+            return;
+        }
+        Dictionary dic = DictionaryUtils.getExternalDictionaryByPayload(mContext, download.getUri());
+        if (dic != null) {
```
Ditto early return

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionaryUtils.java:
```diff
@@ -0,0 +1,206 @@
+    public static String getExternalDicPath(@NonNull Context context, @NonNull String lang) {
+        File file = new File(context.getDatabasePath(getExternalDicFullName(lang)).getAbsolutePath());
+        if (file.exists() && file.isFile()) return file.getAbsolutePath();
+        return null;
```
Mind reversing the logic here? As the method is getExternalDicPath I think it's better if we return null if `!file.exists() || !file.isFile()` and then after that an unconditional `return file.getAbsolutePath()`. It's mainly an aesthetic change but helps to quickly identify what's the expected/default behaviour

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionaryUtils.java:
```diff
@@ -0,0 +1,206 @@
+        if (isBuiltinDictionary(lang)) {
+            return BUILTIN_DICS_MAP.get(lang);
+        }
+        return new String[]{};
```
Same here, let's return an empty String array if there is no builtin dictionary the the builtin one otherwise

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionaryUtils.java:
```diff
@@ -0,0 +1,206 @@
+    @Nullable
+    public static Dictionary getExternalDictionaryByLang(@NonNull Context context, @NonNull String lang, @NonNull String versionName) {
+        Map<String, RemoteProperties> properties = SettingsStore.getInstance(context).getRemoteProperties();
+        if (properties != null) {
```
Early return for null properties?

svillar commented on app/src/common/shared/com/igalia/wolvic/utils/DictionaryUtils.java:
```diff
@@ -0,0 +1,206 @@
+    @Nullable
+    public static Dictionary getExternalDictionaryByPayload(@NonNull Context context, @NonNull String payloadUrl, @NonNull String versionName) {
+        Map<String, RemoteProperties> properties = SettingsStore.getInstance(context).getRemoteProperties();
+        if (properties != null) {
```
Ditto

svillar reviewed (changes requested):
Looks really great! Nice work. As usual :) I have some comments before landing

svillar reviewed (changes requested):
Looks great.

However while testing it I found an issue. The dictionary seems to be only downloaded once you switch keyboards, this means that the dictionary is not downloaded for the default selected keyboard. It's easy to test, my device is in English, and the English dict is not downloaded by default. I switched to spanish and the dictionary was downloaded perfectly. Then I switched to English and this time I got the English completions.

HollowMan6 commented:
> However while testing it I found an issue. The dictionary seems to be only downloaded once you switch keyboards, this means that the dictionary is not downloaded for the default selected keyboard. It's easy to test, my device is in English, and the English dict is not downloaded by default. I switched to spanish and the dictionary was downloaded perfectly. Then I switched to English and this time I got the English completions.

Ah, yes. I was using Chinese in my headset so I completely ignored this one. Sorry about this, will work on a solution for the default keyboard.

HollowMan6 commented:
> > However while testing it I found an issue. The dictionary seems to be only downloaded once you switch keyboards, this means that the dictionary is not downloaded for the default selected keyboard. It's easy to test, my device is in English, and the English dict is not downloaded by default. I switched to spanish and the dictionary was downloaded perfectly. Then I switched to English and this time I got the English completions.
> 
> Ah, yes. I was using Chinese in my headset so I completely ignored this one. Sorry about this, will work on a solution for the default keyboard.

This is now fixed by adding a listener for remote properties: https://github.com/Igalia/wolvic/compare/b171f63bbd7463055db79f8bec7f42554e7c6816..277596e94137c4c813e2062f1dccf139444f7cf6 

This only happens at the first fresh launch of the app, where the remote props haven't been loaded into the settings store. The original PR still downloads dictionaries for the default selected keyboard (after the first launch, reopen Wolvic).

svillar reviewed (approved):
Great work. I'll land this.

If you feel like doing it, a nice followup would be to add some setting to enable/disable it because depending on the device, it can be a bit sluggish making the experience of typing worst.

HollowMan6 commented:
> If you feel like doing it, a nice followup would be to add some setting to enable/disable it because depending on the device, it can be a bit sluggish making the experience of typing worst.

Sure! Actually, it has already been done in the past https://github.com/Igalia/wolvic/pull/1104#issuecomment-1825939500

I will open another PR to land https://github.com/Igalia/wolvic/commit/168ae4138fc9a78ff2c79237477fd4f89ff85809 tonight.
